### ISSUE TITLE
[SEGV] nginx static-linked OpenSSL and ngx_mruby

### ISSUE THREAD

**Issue by cubicdaiya** [matsumoto-r/ngx_mruby#165]
Hi.

If nginx static-linked OpenSSL and ngx_mruby are built, it is possible to be caused SEGV.

## How to reproduce

It is assumed to be already installed OpenSSL as the package.

```
$ wget https://www.openssl.org/source/openssl-1.0.2g.tar.gz
$ wget http://nginx.org/download/nginx-1.9.14.tar.gz
$ tar zxvf openssl-1.0.2g.tar.gz
$ tar zxvf nginx-1.9.14.tar.gz
$ git clone https://github.com/matsumoto-r/ngx_mruby.git
$ cd ngx_mruby
$ ./configure --with-ngx-src-root=../nginx-1.9.14
$ make build_mruby
$ make generate_gems_config
$ cd ../nginx-1.9.14
$ ./configure                     \
 --with-openssl=../openssl-1.0.2g \
 --with-http_ssl_module           \
 --add-module=../ngx_mruby
$ make
$ sudo make install
$ sudo /usr/local/nginx/sbin/nginx -g 'daemon off;'
$ echo $? # SEGV
139
```

Maybe the phenomenon below is similar. (Japanese only, sorry.)

https://twitter.com/d6rkaiz/status/685835026890555392

The patch below is the workaround for this problem. It works fine if it is applied.

```diff
diff --git a/build_config.rb b/build_config.rb
index 46d8941..c4d4ced 100644
--- a/build_config.rb
+++ b/build_config.rb
@@ -10,7 +10,7 @@ MRuby::Build.new do |conf|
   conf.gem :github => 'iij/mruby-io'
   conf.gem :github => 'iij/mruby-env'
   conf.gem :github => 'iij/mruby-dir'
-  conf.gem :github => 'iij/mruby-digest'
+  #conf.gem :github => 'iij/mruby-digest'
   conf.gem :github => 'iij/mruby-process'
   conf.gem :github => 'iij/mruby-pack'
   conf.gem :github => 'iij/mruby-socket'
```

The cause seems to be the OpenSSL dependency for mruby-digest bundled with ngx_mruby by default. mruby-digest requires OpenSSL and it is dynamic-linked when OpenSSL library already exists on system.

On the other hand, nginx provides the mechanism for linking OpenSSL with statically or dynamically.
So if nginx static-linked OpenSSL and ngx_mruby are built, two OpenSSLs seem to conflict.

Thanks.

**Comment by matsumoto-r** [comment]
@cubicdaiya Thank you for your report. I already investigate it at #160 . Did you mean that issue #160 ?

**Comment by cubicdaiya** [comment]
> Did you mean that issue #160 ?

@matsumoto-r Yes. It's same phenomenon.

**Comment by matsumoto-r** [comment]
Thanks! I'll investigate more user-friendley way.

**Comment by ianchen06** [comment]
Hello,
I can confirm the same problem with openssl-1.0.2g install at /usr/local/openssl-1.0.2g
Is there currently a patch available to be applied to mruby-digest for linking with custom openssl?

Thanks

**Comment by matsumoto-r** [comment]
@ianchen06 You should edit `ngx_mruby/mruby/build/mrbgems/mruby-digest/mrbgem.rake` for pointing to ` /usr/local/openssl-1.0.2g` and build again.

For example:

```ruby
  conf.cc do |cc|
    cc.include_paths << '/usr/local/openssl-1.0.2g/include'
  end

  conf.linker do |linker|
    linker.library_paths << '/usr/local/openssl-1.0.2g/lib'
  end
```

**Comment by matsumoto-r** [comment]
@cubicdaiya https://github.com/matsumoto-r/ngx_mruby/issues/165#issuecomment-211759667 What do you think this? This way don't use `--with-openssl=../openssl-1.0.2g` as nginx configuration flag and should build openssl into somewhere in advance.

**Comment by ianchen06** [comment]
@matsumoto-r  Thank you so much for your reply, however I'm getting an error at the LD stage of the build.

I changed `ngx_mruby/mruby/build/mrbgems/mruby-digest/mrbgem.rake` to 
```
MRuby::Gem::Specification.new('mruby-digest') do |spec|
  spec.license = 'MIT'
  spec.authors = 'Internet Initiative Japan Inc.'

  #spec.linker.libraries << 'crypto' unless RUBY_PLATFORM =~ /darwin/
  spec.cc.include_paths << "/usr/local/openssl-1.0.2g/include"
  spec.linker.library_paths << "/usr/local/openssl-1.0.2g/lib"
end
```

and the build fails at

```
$ make build_mruby
...
AR    build/host/lib/libmruby.a
CC    mrbgems/mruby-bin-mirb/tools/mirb/mirb.c -> build/host/mrbgems/mruby-bin-mirb/tools/mirb/mirb.o
LD    build/host/bin/mirb
/home/ian/Code/ngx_mruby/mruby/build/host/lib/libmruby.a(digest.o): In function `lib_hmac_update':
/home/ian/Code/ngx_mruby/mruby/build/mrbgems/mruby-digest/src/digest.c:272: undefined reference to `HMAC_Update'
/home/ian/Code/ngx_mruby/mruby/build/host/lib/libmruby.a(digest.o): In function `lib_hmac_free':
/home/ian/Code/ngx_mruby/mruby/build/mrbgems/mruby-digest/src/digest.c:119: undefined reference to `HMAC_CTX_cleanup'
/home/ian/Code/ngx_mruby/mruby/build/host/lib/libmruby.a(digest.o): In function `lib_hmac_digest_length':
/home/ian/Code/ngx_mruby/mruby/build/mrbgems/mruby-digest/src/digest.c:248: undefined reference to `EVP_MD_size
...
/home/ian/Code/ngx_mruby/mruby/build/mrbgems/mruby-digest/src/digest.c:176: undefined reference to `EVP_sha1'
/home/ian/Code/ngx_mruby/mruby/build/mrbgems/mruby-digest/src/digest.c:178: undefined reference to `EVP_sha256'
collect2: error: ld returned 1 exit status
rake aborted!
Command failed with status (1): [gcc -L"/usr/local/openssl-1.0.2g/lib" -o "...]
/home/ian/Code/ngx_mruby/mruby/tasks/mruby_build_commands.rake:31:in `_run'
/home/ian/Code/ngx_mruby/mruby/tasks/mruby_build_commands.rake:36:in `rescue in _run'
/home/ian/Code/ngx_mruby/mruby/tasks/mruby_build_commands.rake:32:in `_run'
/home/ian/Code/ngx_mruby/mruby/tasks/mruby_build_commands.rake:182:in `run'
/home/ian/Code/ngx_mruby/mruby/Rakefile:70:in `block (4 levels) in <top (required)>'
Command failed with status (1): ["gcc" -L"/usr/local/openssl-1.0.2g/lib" -o...]
/home/ian/Code/ngx_mruby/mruby/tasks/mruby_build_commands.rake:33:in `_run'
```

Am I missing something here, sorry for the newbie questions. Thank you very much for your help

**Comment by matsumoto-r** [comment]
@ianchen06 You're welcome!! delete commented from `#spec.linker.libraries << 'crypto' unless RUBY_PLATFORM =~ /darwin/` and try build again.

**Comment by cubicdaiya** [comment]
@matsumoto-r 

IMO, the way above is the workaround and not user-friendly. So it is better that ngx_mruby provides the transparent and successful build-way even if `--with-openssl=PATH` is assigned. (At least, by default configuration.) Because `--with-openssl=PATH` is the feature nginx-self provides and I think it is maybe the susceptible pitfall for a user embeds mruby into nginx.

**Comment by matsumoto-r** [comment]
@cubicdaiya 

> the way above is the workaround and not user-friendly

I agree with you. So, I'll consider steps to build custom openssl transparently like `./configure --with-ngx-src-root=../nginx-1.9.14 --with-openssl==../openssl-1.0.2g` as ngx_mruby configure option. Thank you for your opinion.

**Comment by ianchen06** [comment]
@matsumoto-r 
It is working now, thank you very much for your help.

**Comment by matsumoto-r** [comment]
Fixed via #167 .
